### PR TITLE
Posdis metric as a callback

### PR DESCRIPTION
**Pull request by eugene-kharitonov**

### CODE REVIEW

File: egg/core/language_analysis.py [hunk 1]
```diff
@@ -28,7 +27,11 @@ def entropy_dict(freq_table):
     True
     """
     t = torch.tensor([v for v in freq_table.values()]).float()
-    return torch.distributions.Categorical(probs=t).entropy().item() / np.log(2)
+    if (t < 0.0).any():
+        raise RuntimeError('Encountered negative probabilities')
+
+    t /= t.sum()
+    return -(torch.where(t > 0, t.log(), t) * t).sum().item() / np.log(2)
```
Comment (eugene-kharitonov) [inline egg/core/language_analysis.py:34]: Silly Categorical doesn't allow prob=0, clamping them at some small value

File: egg/core/language_analysis.py [hunk 2]
```diff
@@ -47,10 +50,32 @@ def calc_entropy(messages):
 
 
 def _hashable_tensor(t):
-    t = tuple(t.tolist())
+    if torch.is_tensor(t) and t.numel() > 1:
+        t = tuple(t.tolist())
+    elif torch.is_tensor(t) and t.numel() == 1:
```
Comment (eugene-kharitonov) [inline egg/core/language_analysis.py:55]: `torch.tensor(80).tolist()` would return a single number...

File: egg/core/language_analysis.py (hunk 1 above)
Comment (eugene-kharitonov) [reply egg/core/language_analysis.py:34]: yeah, but 
```
>>> torch.distributions.Categorical(probs=torch.tensor([0, 1.0])).entropy().item() / np.log(2)
1.7198267136473827e-07
```
which is positive...

Comment (robertodessi) [reply egg/core/language_analysis.py:34]: how about `round(torch.distributions.Categorical(probs=torch.tensor([0.0, 1.0])).entropy().item() / np.log(2), 4)`

**Review by robertodessi — APPROVED** [review]

File: egg/core/language_analysis.py [hunk 3]
```diff
@@ -124,3 +149,70 @@ def compute_distance(_list, distance):
...
+
+
+class PosDisent(Callback):
+    """
+    Positional disentanglement metric, introduced in "Compositionality and Generalization in Emergent Languages", 
+    Chaabouni et al., ACL 2020.
+    """
+    def __init__(self, print_train: bool = True, is_gumbel: bool = False):
+        super().__init__()
+        self.print_train = print_train
```
Comment (robertodessi) [inline egg/core/language_analysis.py:161]: I've found computing posdis on the train set to slow down training quite a lot for the multitask multiagent game. In genera, wouldn't it be better to have it computed on the validation data only by default?

File: egg/core/language_analysis.py [hunk 4]
```diff
@@ -124,3 +149,70 @@ def compute_distance(_list, distance):
...
+
+        output = json.dumps(dict(posdis=posdis, mode=tag, epoch=epoch))
+        print(output, flush=True)
+
+    def on_epoch_end(self, _loss, logs: Interaction, epoch: int):
+        if self.print_train:
+            self.print_message(logs, 'train', epoch)
+
+    def on_test_end(self, loss, logs, epoch):
+        self.print_message(logs, 'test', epoch)
```
Comment (robertodessi) [inline egg/core/language_analysis.py:218]: In the [topsim](https://github.com/facebookresearch/EGG/blob/master/egg/core/language_analysis.py#L103-L111) callback I made both (print_train and print_valid) optional but at least one must be specified. 



I'd say to implement both callbacks with the same feature for consistency. Both are fine with me

File: egg/core/language_analysis.py (hunk 1 above)
Comment (eugene-kharitonov) [reply egg/core/language_analysis.py:34]: looks more magic to me than the original version =D

File: egg/core/language_analysis.py (hunk 3 above)
Comment (eugene-kharitonov) [reply egg/core/language_analysis.py]: Done

File: egg/core/language_analysis.py (hunk 4 above)
Comment (eugene-kharitonov) [reply egg/core/language_analysis.py]: Done
